security/pkg/nodeagent/sds/monitoring.go

```diff
@@ -0,0 +1,164 @@
+		prometheus.GaugeOpts{
+			Namespace: "citadel_agent",
+			Name:      "pushed_root_cert_expiry_timestamp",
+			Subsystem: "sds_service",
+			Help:      "The date after which a pushed root certificate expires. Expressed as a Unix Epoch Time.",
+		}, []string{"resourcePerConn"})
+
+	// serverCertExpiryTimestamp records the expiration timestamp of the most recent pushed server
+	// certificate for a particular SDS resource. The label of a pushed root cert is represented as
+	// <resource name>-<connection ID>, and the value is in Unix Epoch Time.
+	serverCertExpiryTimestamp = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "citadel_agent",
+			Name:      "pushed_server_cert_expiry_timestamp",
+			Subsystem: "sds_service",
+			Help:      "The date after which a pushed server certificate expires. Expressed as a Unix Epoch Time.",
+		}, []string{"resourcePerConn"})
+)
+
+func init() {
+	prometheus.MustRegister(totalPushCounts)
+	prometheus.MustRegister(totalPushErrorCounts)
+	prometheus.MustRegister(totalActiveConnCounts)
+	prometheus.MustRegister(totalStaleConnCounts)
+	prometheus.MustRegister(pendingPushPerConnCounts)
+	prometheus.MustRegister(staleConnCounts)
+	prometheus.MustRegister(pushPerConnCounts)
+	prometheus.MustRegister(rootCertExpiryTimestamp)
+	prometheus.MustRegister(serverCertExpiryTimestamp)
+}
+
+// monitoringMetrics are counters for SDS push related operations.
+type monitoringMetrics struct {
+	totalPush                 prometheus.Counter
+	totalPushError            prometheus.Counter
+	totalActiveConn           prometheus.Gauge
+	totalStaleConn            prometheus.Gauge
+	pendingPushPerConn        *prometheus.GaugeVec
+	staleConn                 *prometheus.CounterVec
+	pushPerConn               *prometheus.CounterVec
+	pushErrorPerConn          *prometheus.CounterVec
+	rootCertExpiryTimestamp   *prometheus.GaugeVec
+	serverCertExpiryTimestamp *prometheus.GaugeVec
+}
+
+// newMonitoringMetrics creates a new monitoringMetrics.
+func newMonitoringMetrics() monitoringMetrics {
+	return monitoringMetrics{
+		totalPush:                 totalPushCounts,
+		totalPushError:            totalPushErrorCounts,
+		totalActiveConn:           totalActiveConnCounts,
+		totalStaleConn:            totalStaleConnCounts,
+		pendingPushPerConn:        pendingPushPerConnCounts,
+		staleConn:                 staleConnCounts,
+		pushPerConn:               pushPerConnCounts,
+		pushErrorPerConn:          pushErrorsPerConnCounts,
+		rootCertExpiryTimestamp:   rootCertExpiryTimestamp,
+		serverCertExpiryTimestamp: serverCertExpiryTimestamp,
+	}
+}
+
+// generateResourcePerConnLabel returns a label that can be used to differentiate metrics by
+// resource name and connection ID.
+// An example label is
+// bookinfo-credential-1-cacert+router~10.60.4.32~istio-ingressgateway-5b4458d75b-5kqhg.istio-system~istio-system.svc.cluster.local-7
+func generateResourcePerConnLabel(resourceName, conID string) string {
```

Comment:
This is a rather long label. We have had issues with long labels and high cardinality before.

Reply:
These metrics with labels will not be sent to Prometheus, we can use /metrics endpoint to monitor these metrics. Is that still going to cause any issues?